cli/connhelper/commandconn/commandconn.go

```diff
@@ -113,24 +111,32 @@ func (c *commandConn) kill() error {
 		werr = c.cmd.Wait()
 	}
 	c.cmdWaitErr = werr
-	c.cmdExited = true
 	c.cmdMutex.Unlock()
+	c.cmdExited.Store(true)
 	return nil
 }
 
+// onEOF gets called if we receive an io.EOF while reading
+// or writing from the undelying command pipes.
+//
+// When we've received an EOF we expect that the command will
+// be terminated soon. As such, we call Wait() on the command
+// and return EOF or the error depending on whether the command
+// exited with an error.
+//
+// If Wait() does not return within 10s, an error is returned
 func (c *commandConn) onEOF(eof error) error {
-	// when we got EOF, the command is going to be terminated
 	var werr error
 	c.cmdMutex.Lock()
```

Comment:
Curious now; should we keep the lock for the whole function, or do we only need it to protect `c.cmdWaitErr` ?

In the latter case, we could change these to

```go
c.cmdMutex.Lock()
werr = c.cmdWaitErr
c.cmdMutex.Unlock()
```

and (second "if branch")

```go
c.cmdMutex.Lock()
c.cmdWaitErr = werr
c.cmdMutex.Unlock()
```

Reply:
As an aside, this discussion really makes me think this whole code could use a big refactor. I started one while I was digging around and got a fair bit into it, but then realized I had already wasted too much time on it and went back to my previous branch and just applied these few changes. Maybe it's worth it though, I'm definitely keeping this in the back pocket for when I'm a bit bored (hah)